model/trace/trace-exception.yaml

```diff
@@ -7,10 +7,10 @@ groups:
       report a single exception associated with a span.
     attributes:
       - ref: exception.type
+        requirement_level:
+          conditionally_required: Required if `exception.message` is not set, recommended otherwise.
       - ref: exception.message
+        requirement_level:
+          conditionally_required: Required if `exception.type` is not set, recommended otherwise.
       - ref: exception.stacktrace
       - ref: exception.escaped
-    constraints:
-      - any_of:
-          - "exception.type"
-          - "exception.message"
```

Comment:
One benefit of this would be that it could be automatically verified/enforced whereas the prose equivalent is only accessible to humans.

@lquerel do you make any use of these constraints in https://github.com/open-telemetry/weaver?

Reply:
I don't think so (see [this Slack thread](https://cloud-native.slack.com/archives/C0697EXNTL3/p1710780112791009?thread_ts=1710727871.932029&cid=C0697EXNTL3)), but it would still be good for @lquerel to weigh in.

I'm fine with removing constraints, as they try to model a cross-attribute requirement level, which doesn't play nicely with the attribute-specific requirement level.

Reply:
>  as they try to model a cross-attribute requirement level, which doesn't play nicely with the attribute-specific requirement level

What is the non-nice thing exactly? Maybe something that can be addressed by generating the markdown differently?

Reply:
the context - https://github.com/open-telemetry/build-tools/issues/293

Reply:
The markdown (hiding requirement level are replacing it with `See below`) can be fixed, but constraints create the pattern that you can see in the https://github.com/open-telemetry/semantic-conventions/pull/863

where we can't decide on single common attribute and create a subpar experience for users. 
I argue that exceptions semconv would be better if we agreed on specific attribute that's required in all cases.

Reply:
I think the problem here is that you view the semantic conventions as a specification for what you have available at query-time while when we implemented constraints, we were rather thinking of specifying what should be reported by instrumentations, so that then some normalization component (backend, collector, ...) could normalize to something that is easier to query.

I.e., the key question is: Are the semantic conventions, especially requirements, binding for instrumentation authors, or are they meant as the final requirements of what the data you have stored in the backend should look like, no matter where in the pipeline they are filled in.

Reply:
I believe we went away from "collect what's available, normalize on the backend" everywhere except exceptions and process semconv. 
Instrumentation is the best (and sometimes the only) place which knows how to populate certain data.
With HTTP we saw that instrumentation authors now knowing what to et, set everything, we clearly see the same pattern with exceptions (both message and type are recorded):

- [Java](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/internal/data/ImmutableExceptionEventData.java)
- [Python](https://github.com/open-telemetry/opentelemetry-python/blob/05497a53051c9b2d93ec1da9d3284b7c9cff830f/opentelemetry-sdk/src/opentelemetry/sdk/_logs/_internal/__init__.py#L469)
- [Go](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/span.go#L431)

Reply:
> What is the non-nice thing exactly? Maybe something that can be addressed by generating the markdown differently?

There's not restriction of what attributes I put in an `any_of` list, and to me it's not conceptually clear what we mean by `any_of` two attributes that are optional, or `any_of` two attributes that are optional, or `any_of` two attributes where one is required and one is optional.

In case of `any_of` two  attributes that are required, I assume it means that only one of the two attributes is required. However, doesn't that make the attributes "conditionally required" (required on the condition that the other attribute is absent)?

I think removing `any_of` removes a layer of complexity that can potentially cause ambiguity.

Reply:
resolving this thread based on https://github.com/open-telemetry/semantic-conventions/pull/862#issuecomment-2033190019